### DISCUSSION TITLE
Better default bindings for exiting insert mode

### DISCUSSION THREAD

**Discussion by VisenDev** [helix-editor/helix#11710] [Ideas]
When experimenting with the default keybindings for helix, I was disappointed to find that `<Esc>` seemed to be the only builtin keybinding for exiting insert mode. 

Due to the inconvenient placement of the escape key on most modern keyboards, almost all users of modal editors need to add a more ergonomic keybinding for `<Esc>` to their configuration. Whenever I use vim on an unfamiliar system, I always have to run `:imap jk <Esc>` to be productive.

Since one of the focuses of Helix is modern sensible defaults, it might be worthwhile to add some of the more common keybindings for exiting insert mode to helix. For example, `jk` or`ii`

Vim also supports using ctrl-c to exit insert mode by default, which I find quicker to use than escape.

`jk` is my personal favorite way to exit insert mode, but anything that is more convenient than `<Esc>` would be nice to have as a default binding.

**Comment by David-Else** [comment]
I thought exactly the same as you when I first got in to modal editing. I discovered later that making caps lock escape on tap and ctrl on hold was by far the best way to do it, and it works great on the rest of the operating system too.

If you are on Linux you can remap with https://github.com/rvaiya/keyd

**Reply by VisenDev** [reply]
I am aware such remappings exist (I have already set my keyboard up this way)

However, requiring custom keyboard configuration to use the editor efficiently defeats the purpose of *usable defaults*. Ideally, I could freshly install helix on a new machine and be able to use it without needing to remap my keyboard or write configuration files.

**Reply by David-Else** [reply]
There really is only 2 common defaults for modal editors to exit insert mode:

1. For beginners who have not modded their keyboard, JK
2. For experienced users with modded keys or special keyboards, ESC.

I think adding JK as a default would annoy experienced users as they might press it by accident. I would support adding some text to the docs to clearly explain about keyboard bindings and suggest JK as an option.

**Reply by VisenDev** [reply]
> 1. For beginners who have not modded their keyboard, JK
> 2. For experienced users with modded keys or special keyboards, ESC.

I disagree with this characterization of beginners vs experienced user's solutions to this problem. I have years of experience using modal editors and still prefer using bindings like `jk` instead of remapping caps lock because
- `jk` is much more ergonomic to type because it uses my strongest two fingers on my strongest hand while at home position, versus `capslock` requires me to use my weakest finger on my weakest hand and I must move the finger away from home position.
- For non-linux users or for users who are not administrators on the computer they are working on, remapping caps lock is not a realistic option. 

> I think adding JK as a default would annoy experienced users as they might press it by accident.

I think it is more appropriate to expect experienced users to disable something unwanted in their configuration file than it is to require new users to edit their configuration file to enable something most users will want. (Though I could understand certain bindings being annoying)

Edit:
I'd also add that though `jk` is my preferred way to exit insert mode, I'm not arguing for this binding specifically. Almost any default binding other than `<Esc>` would be sufficient.  At the very least, ctrl-c to exit insert mode (as in vim and neovim) should be enabled in my opinion

**Comment by rhogenson** [comment]
I agree with this, it's a bit unfortunate to ask all users to rebind their keyboard at the OS level to get ergonomic keybinds. I tried Helix because I heard it has "good defaults"

**Comment by rahil627** [comment]
jk is _very_ opinionated. i tried it when i first discovered it in an example in the docs, and absolutely loathed it. sooooooo annoying! Insert mode is for inserting characters, and letters are the most important keys!

a control-key shortcut however sounds fine, less bothersome, as it won't conflict with other modes.

when i saw the topic, i was actually hoping for more sensible default key-mappings in insert-mode. it's mappings are so sparse..!

maybe control+enter or shift+enter, to have a big key, and to provoke the ancient feeling of pressing (carriage-)return on a typewriter? 😄 

people often bind control+q to just straight `:quit`, in normal-mode, but still... control+c is used to (c)lose in picker and prompt, so that sounds good!

**Reply by VisenDev** [reply]
> more sensible default key-mappings in insert-mode

Do you mean emacs style text editing controls? (like ctrl-a to go to the beginning of the line, ctrl-e to go to the end, etc...?)

**Reply by rahil627** [reply]
precisely! just making insert-mode more useful... tho, i can see how it goes against the ancient traditional ways of vi.. i really loved having the helix plugin for vs-code, a sort of nasty mish-mash of both key-bindings... it eased the transition as i could always fallback on using the basic controls. i tend to feel like vi is more useful for _editing_, whereas vs-code (emacs?) is more useful for _writing_. and we all should be doing more writing..!

(note: i haven't used emacs... _yet_. i think i'll try lem first, then doom. but i do believe lisp is indeed the ultimate way to go, and not just limited to scripting.)

in my config i even added this early on, just so i can be immediately productive:
```
C-left = "move_prev_word_start"
C-right = "move_next_word_end"
```
it probably should be c-w/c-e/c-b, but you get the point: there's no movement at all in there!!

**Comment by markstos** [comment]
>  Whenever I use vim on an unfamiliar system, I always have to run :imap jk <Esc> to be productive.

I use and recommend the "special keyboard" method described above. With I keyboard running QMK firmware, I've configured it so holding JK together for 40 ms-- in any app-- sends the Escape key. I have the Escape key mapped somewhere else on my keyboard, but in practice, the JK combo is so convenient I use it all the time.

**Reply by rahil627** [reply]
LOL :sob:. this got me dyin'. vi just consumes people. first the editor, then terminal 'n terminal apps including tmux 'n display manager, soon the OS, later real life. if only life's interface was a keyboard.

**Comment by cassepipe** [comment]
/begin rant

It's always the same when someone brings up  this problem, you always have an army of people that explains either how *it's fine actually* and you can get it used to it (only a third a of vimmers use an Esc alternative according to reddit polls) or how they fixed it with their clever solution (even though with time it seems that a consensus on solutions emerged)

But everyone is *missing the point*, the question at hand is *bad defaults*. One can fix or get used to anything but it creates unecessary friction, yet another thing to configure across your machines. 

And it's not like there is a good reason for this, it's a historical accident !
Even the documentation of the different modal editors don't document  the most famous workarounds.

Bad defaults are the bane of existence and people who defend them are the reason we can't have nice things.

/end rant

Edit: Well at least there is  [this](https://github.com/helix-editor/helix/wiki/FAQ#map-jk-or-jj-to-exit-insert-mode) in Helix's wiki but still this is like the barest of bare minimums.

**Comment by Qcucchiara** [comment]
As an absolute beginner
esc: too far (and I will not start customizing my keyboard for that, it's crazy)
jk: is fine, but a bit annoying when I actually want to type just a "j" letter in insert

I just tried A-i (alt + i) and it seem to work.

**Reply by cassepipe** [reply]
It's kind of the issue, either people suffer the inconvenience to be part of the cult of the cool kids, or they stumble on someone suggesting a workaround (around 1/3 of vimmers use a workaround according to reddit polls, I can't believe it's that low)

I settled on mapping CapsLock to Escape because it's just a gui settings under linuxes and MacOs, and on Windows you just have to modify a register key, so it can be done quickly on any machine.

Another advantage is that it will also work for vi modes in other software for free (zsh, bash, fish, gdb, etc. , also not for the firefox console, you have to use Ctrl+C)

Also you will find that it's quite nice to be able to escape any annoying situtation by pressing CapsLock

**Comment by rahil627** [comment]
yeah, re-binding caps-lock at OS level is probably the most common solution...

how about making the default binding of exit-mode mapped to the actual caps lock (in insert-mode)? then it should work for both camps: newbies and the sane users. then maybe can hide the caps-lock command in control-caps-lock/esc or alt-caps-lock/esc or something (in insert-mode)..
